### PR TITLE
[FLINK-30247] Introduce time travel reading for table store

### PR DESCRIPTION
**Pull request by zjureel**
Introduce time travel reading for table store

### CODE REVIEW

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/AbstractFileStoreScan.java [hunk 1]
```diff
@@ -184,10 +203,14 @@ public Plan plan() {
...
-                        readCompacted
-                                ? snapshotManager.latestCompactedSnapshotId()
-                                : snapshotManager.latestSnapshotId();
+                if (timestampMills == null) {
+                    snapshotId =
+                            readCompacted
+                                    ? snapshotManager.latestCompactedSnapshotId()
+                                    : snapshotManager.latestSnapshotId();
+                } else {
+                    snapshotId = snapshotManager.earlierThanTimeMills(timestampMills);
```
Comment (JingsongLi) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/AbstractFileStoreScan.java:212]: Maybe we should use `earlier Than or equals`?

**Review by JingsongLi — COMMENTED** [review]
Thanks @zjureel for the contribution, looks nice!
I think we can take a look to https://issues.apache.org/jira/browse/FLINK-30293 , maybe it is better to extract a enumunator for batch too. Keep Scan simple and clean.

What do you think?

**Comment by zjureel** [comment]
> 

Pretty good! I will fix https://issues.apache.org/jira/browse/FLINK-30293 first, please assign it to me, THX

File: docs/layouts/shortcodes/generated/core_configuration.html [hunk 2]
```diff
@@ -8,6 +8,18 @@
...
     </thead>
     <tbody>
+        <tr>
+            <td><h5>as-of-snapshot</h5></td>
+            <td style="word-wrap: break-word;">(none)</td>
+            <td>Long</td>
+            <td>Read snapshot specific by snapshot id.</td>
+        </tr>
+        <tr>
+            <td><h5>as-of-timestamp-mills</h5></td>
```
Comment (tsreaper) [inline docs/layouts/shortcodes/generated/core_configuration.html:18]: Why introduce a new configuration? Time travel reading is actually very similar to the "from-timestamp" startup mode in streaming jobs. I'm considering supporting startup mode for both batch and streaming jobs. See https://issues.apache.org/jira/browse/FLINK-30294 .

**Comment by zjureel** [comment]
Hi @JingsongLi @tsreaper  I have updated the codes based on the master branch, please help to review again

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/snapshot/FromTimestampStartingScanner.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+    public FromTimestampStartingScanner(long startupMillis) {
+        this.startupMillis = startupMillis;
+    }
+
+    @Override
+    public DataTableScan.DataFilePlan getPlan(SnapshotManager snapshotManager, DataTableScan scan) {
+        Long startingSnapshotId = snapshotManager.earlierOrEqualTimeMills(startupMillis);
+        if (startingSnapshotId == null) {
+            LOG.debug("There is currently no snapshot. Waiting for snapshot generation.");
```
Comment (tsreaper) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/snapshot/FromTimestampStartingScanner.java:47]: Batch jobs do not wait for snapshot generation. They just read an empty table.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/snapshot/FromSnapshotStartingScanner.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+import org.apache.flink.table.store.file.operation.ScanKind;
+import org.apache.flink.table.store.file.utils.SnapshotManager;
+import org.apache.flink.table.store.table.source.DataTableScan;
+
+/**
+ * {@link StartingScanner} for the {@link
+ * org.apache.flink.table.store.CoreOptions.StartupMode#FROM_SNAPSHOT} startup mode of a batch read.
+ */
+public class FromSnapshotStartingScanner implements StartingScanner {
```
Comment (tsreaper) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/snapshot/FromSnapshotStartingScanner.java:29]: `StaticFromSnapshotStartingScanner`.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/snapshot/FromTimestampStartingScanner.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link StartingScanner} for the {@link
+ * org.apache.flink.table.store.CoreOptions.StartupMode#FROM_TIMESTAMP} startup mode of a batch
+ * read.
+ */
+public class FromTimestampStartingScanner implements StartingScanner {
```
Comment (tsreaper) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/snapshot/FromTimestampStartingScanner.java:33]: `StaticFromTimestampStartingScanner`.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/snapshot/ContinuousFromSnapshotStartingScanner.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+    private final long snapshotId;
+
+    public ContinuousFromSnapshotStartingScanner(long snapshotId) {
+        this.snapshotId = snapshotId;
+    }
+
+    @Override
+    public DataTableScan.DataFilePlan getPlan(SnapshotManager snapshotManager, DataTableScan scan) {
+        return new DataTableScan.DataFilePlan(snapshotId, Collections.emptyList());
```
Comment (tsreaper) [inline flink-table-store-core/src/main/java/org/apache/flink/table/store/table/source/snapshot/ContinuousFromSnapshotStartingScanner.java:47]: `from-timestamp` includes the changes at that timestamp, but `from-snapshot` does not include the changes at that snapshot? Why do you choose this behavior, please clarify.

File: docs/content/docs/how-to/querying-tables.md [hunk 7]
```diff
@@ -180,3 +180,13 @@ SELECT * FROM MyTable$audit_log;
...
+
+## Time Travel Reading
+You can read snapshot specific by commit time or snapshot id from a table.
+```sql
+-- Read snapshot specified by commit time.
+SELECT * FROM T /*+ OPTIONS('scan.timestamp-millis'='121230')*/;
+
+-- Read snapshot specified by snapshot id.
+SELECT * FROM T /*+ OPTIONS('scan.snapshot-id'='12')*/;
+```
```
Comment (tsreaper) [inline docs/content/docs/how-to/querying-tables.md:192]: No need to create a separate section, just update the table in "Scan Mode" section.

Comment (zjureel) [reply docs/content/docs/how-to/querying-tables.md]: Done

**Review by tsreaper — APPROVED** [review]

**Pull request merged**